### PR TITLE
VITIS-12792 HIP Binding: Memory pool APIs

### PR DESCRIPTION
**Pull request by zhangchiming**
Problem solved by the commit
add initial implementation of hip stream ordered memory allocator.

How problem was solved, alternative solutions (if any) and why they were rejected
N/A

Risks (if any) associated the changes in the commit
None. Code will only be built with switch "-hip".

What has been tested and how, request additional testing if necessary
Compiled and tested on Ubuntu 22.04 (running on Ryzen 7840).

Documentation impact (if any)
None.

### CODE REVIEW

File: src/runtime_src/hip/api/hip_device.cpp [hunk 1]
```diff
@@ -66,7 +70,7 @@ hip_init(unsigned int flags)
   std::call_once(device_init_flag, xrt::core::hip::device_init);
 }
 
-static size_t
+static int
```
Comment (stsoe) [inline src/runtime_src/hip/api/hip_device.cpp:73]: Why `int`.  Should be unsigned and `size_t` is the proper type to use here.
It's possible hipGetDeviceCount is using `int*`, but then you static_cast.

File: src/runtime_src/hip/api/hip_device.cpp [hunk 2]
```diff
@@ -146,7 +150,7 @@ hipInit(unsigned int flags)
 }
 
 hipError_t
-hipGetDeviceCount(size_t* count)
+hipGetDeviceCount(int* count)
```
Comment (stsoe) [inline src/runtime_src/hip/api/hip_device.cpp:153]: I assume int  is required by HIP?

File: src/runtime_src/hip/api/hip_device.cpp [hunk 3]
```diff
@@ -48,6 +49,9 @@ device_init()
       continue;
     auto dev = std::make_shared<xrt::core::hip::device>(i);
     device_cache.add(i, std::move(dev));
+    auto mem_pool = std::make_shared<xrt::core::hip::memory_pool>(device_cache.get_or_error(i), MAX_MEMORY_POOL_SIZE_NPU, MEMORY_POOL_BLOCK_SIZE_NPU);
```
Comment (stsoe) [inline src/runtime_src/hip/api/hip_device.cpp:52]: So the mem_pool is shared with whom?  I am suspicious about shared ownership, but I do see the mem_pool pushed to a _db and a map, but is there one of these who is the owner and the other one an observer?   In other words do one outlive the other?    

Same comment for device, which seems to be owned by the device_cache, which should then be unique, but again, I am not sure because the code is not commented.

File: src/runtime_src/hip/api/hip_memory.cpp [hunk 4]
```diff
@@ -225,7 +241,7 @@ namespace xrt::core::hip
    
     auto s_hdl = hip_stream.get();
     auto cmd_hdl = insert_in_map(command_cache,
-                                std::make_shared<copy_buffer>(hip_stream, XCL_BO_SYNC_BO_TO_DEVICE, hip_mem_dst, src, size, offset));
+                                 std::make_shared<memcpy_command>(hip_stream, dst, src, size, hipMemcpyHostToDevice));
```
Comment (stsoe) [inline src/runtime_src/hip/api/hip_memory.cpp:244]: shared vs unique?  Some comments would be nice.

**Review by stsoe — CHANGES_REQUESTED** [review]
Please undo the Copyright changes.

In general I am concerned about defaulting to make_shared to avoid explaining who really owns the objects that are created.  make_shared is cool for shared ownership, but it should not be abused as this can lead to a tangled mess of improperly layered runtime architecture.

File: src/runtime_src/hip/api/hip_device.cpp [hunk 5]
```diff
@@ -1,10 +1,11 @@
 // SPDX-License-Identifier: Apache-2.0
-// Copyright (C) 2023-2024 Advanced Micro Devices, Inc. All rights reserved.
+// Copyright (C) 2023-2024 Advanced Micro Device, Inc. All rights reserved.
```
[the remark this answers is not in the document]
Comment (zhangchiming) [reply src/runtime_src/hip/api/hip_device.cpp]: Sorry, rebase error. It has been reverted back.

File: src/runtime_src/hip/api/hip_memory.cpp [hunk 6]
```diff
@@ -1,5 +1,5 @@
 // SPDX-License-Identifier: Apache-2.0
-// Copyright (C) 2023 Advanced Micro Devices, Inc. All rights reserved.
+// Copyright (C) 2023 Advanced Micro Device, Inc. All rights reserved.
```
[the remark this answers is not in the document]
Comment (zhangchiming) [reply src/runtime_src/hip/api/hip_memory.cpp]: Sorry, rebase error. It has been reverted back.

File: src/runtime_src/hip/api/hip_device.cpp (hunk 1 above)
Comment (zhangchiming) [reply src/runtime_src/hip/api/hip_device.cpp]: Sorry, rebase error. It has been reverted back.

File: src/runtime_src/hip/api/hip_device.cpp (hunk 2 above)
Comment (zhangchiming) [reply src/runtime_src/hip/api/hip_device.cpp]: Sorry, rebase error. It has been reverted back.

File: src/runtime_src/hip/api/hip_memory.cpp (hunk 4 above)
Comment (zhangchiming) [reply src/runtime_src/hip/api/hip_memory.cpp:244]: my understanding is that the ptr to a xrt::core::hip::command object could be shared between global command_cache and stream::m_top_event::m_chain_of_commands of a stream object. So shared_ptr is chosen for the storage type.

Comment has been added to the code.

File: src/runtime_src/hip/core/event.cpp [hunk 7]
```diff
@@ -1,5 +1,5 @@
 // SPDX-License-Identifier: Apache-2.0
-// Copyright (C) 2024 Advanced Micro Devices, Inc. All rights reserved.
+// Copyright (C) 2024 Advanced Micro Device, Inc. All rights reserved.
```
Comment (stsoe) [inline src/runtime_src/hip/core/event.cpp:2]: Undo this change please.  It is `Devices` not `Device`

File: src/runtime_src/hip/api/hip_memory.cpp [hunk 8]
```diff
@@ -193,6 +194,22 @@ namespace xrt::core::hip
...
+  static void
+  hip_memcpy_async(void* dst, const void* src, size_t size, hipMemcpyKind kind, hipStream_t stream)
+  {
+    throw_invalid_value_if(!dst, "dst is nullptr.");
+    throw_invalid_value_if(!src, "src is nullptr.");
+
+    auto hip_stream = get_stream(stream);
+    throw_invalid_value_if(!hip_stream, "Invalid stream handle.");
+
+    // ptr to a xrt::core::hip::command object could be shared between global command_cache and stream::m_top_event::m_chain_of_commands of a stream object
```
Comment (stsoe) [inline src/runtime_src/hip/api/hip_memory.cpp:206]: Lets split this into lines approximately 80 chars long

File: src/runtime_src/hip/core/memory_pool.cpp [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+  {
+    if (node->m_prev == nullptr) {
+      head = node->m_next;
+      if (node->m_next != nullptr) node->m_next->m_prev = nullptr;
+    } else {
+      node->m_prev->m_next = node->m_next;
+      if (node->m_next!= nullptr) {
+        node->m_next->m_prev = node->m_prev;
+      }
+    }
```
Comment (stsoe) [inline src/runtime_src/hip/core/memory_pool.cpp:37]:
```suggestion
    if (node->m_prev == nullptr) {
      head = node->m_next;
      if (node->m_next != nullptr) 
        node->m_next->m_prev = nullptr;
    } 
    else {
      node->m_prev->m_next = node->m_next;
      if (node->m_next!= nullptr)
        node->m_next->m_prev = node->m_prev;
    }
```
I think code is easier to read when `else` aligns with `if`.  Especially in this case where this is an embedded if.
Never put statement at same line as condition, which makes debugging hard.

File: src/runtime_src/hip/core/memory_pool.cpp [hunk 10]
```diff
@@ -0,0 +1,407 @@
...
+
+  static void
+  dlinkedlist_insert(std::shared_ptr<memory_pool_slot>& head, std::shared_ptr<memory_pool_slot>& node)
+  {    
+    node->m_prev = nullptr;
+    node->m_next = head;
+    if (head != nullptr) {
+      head->m_prev = node;
+    }
+    head = node;
```
Comment (stsoe) [inline src/runtime_src/hip/core/memory_pool.cpp:48]:
```suggestion
    if (head != nullptr)
      head->m_prev = node;
    
    head = node;
```
Nit.  Just what most code does, no curly braces for single statement, but new line after

File: src/runtime_src/hip/core/memory_pool.cpp [hunk 11]
```diff
@@ -0,0 +1,407 @@
...
+
+  void
+  memory_pool_node::merge_free_slots(std::shared_ptr<memory_pool_slot> new_free_slot)
+  {
+    std::shared_ptr<memory_pool_slot> p0 = new_free_slot;
+    std::shared_ptr<memory_pool_slot> p1 = new_free_slot;
+
+    while (p0->is_free()) {
+      p1 = p0;
+      if (p0->start() == 0) break;
```
Comment (stsoe) [inline src/runtime_src/hip/core/memory_pool.cpp:71]:
```suggestion
      if (p0->start() == 0) 
        break;
        
```

File: src/runtime_src/hip/core/memory_pool.cpp [hunk 12]
```diff
@@ -0,0 +1,407 @@
...
+
+  size_t
+  memory_pool_node::free(size_t start)
+  {
+    size_t size_freed = 0;
+    auto used_slot  = m_alloc_list;
+    while (used_slot != nullptr)
+    {
+      if (used_slot->m_start == start )
+      {
```
Comment (stsoe) [inline src/runtime_src/hip/core/memory_pool.cpp:91]:
```suggestion
      if (used_slot->m_start == start ) {
```
Consistency please.

File: src/runtime_src/hip/core/memory_pool.cpp [hunk 13]
```diff
@@ -0,0 +1,407 @@
...
+      p0 = p0->next();
+    }
+  }
+
+  size_t
+  memory_pool_node::free(size_t start)
+  {
+    size_t size_freed = 0;
+    auto used_slot  = m_alloc_list;
+    while (used_slot != nullptr)
```
Comment (stsoe) [inline src/runtime_src/hip/core/memory_pool.cpp:88]:
```suggestion
    while (used_slot)
```

File: src/runtime_src/hip/core/memory_pool.cpp [hunk 14]
```diff
@@ -0,0 +1,407 @@
...
+        size_freed = used_slot->get_size();
+        m_used -= size_freed;
+        dlinkedlist_delete(m_alloc_list, used_slot);
+        dlinkedlist_insert(m_free_list, used_slot);
+        used_slot->m_is_free = true;
+        merge_free_slots(used_slot);
+        break;
+      }
+      used_slot = used_slot->m_next;
+    };
```
Comment (stsoe) [inline src/runtime_src/hip/core/memory_pool.cpp:101]:
```suggestion
    }
```

File: src/runtime_src/hip/core/memory_pool.cpp [hunk 15]
```diff
@@ -0,0 +1,407 @@
...
+    m_reserved_mem_current = m_pool_size;
+
+    if (m_pool_size > m_max_total_size)
+    {
+      throw std::runtime_error("mem poolsize is too big.");
+    }
+    else if (m_pool_size == m_max_total_size)
+    {
+      m_auto_extend = false;
+    }
```
Comment (stsoe) [inline src/runtime_src/hip/core/memory_pool.cpp:127]:
```suggestion
    if (m_pool_size > m_max_total_size)
      throw std::runtime_error("mem poolsize is too big.");
      
    if (m_pool_size == m_max_total_size)
      m_auto_extend = false;
```

File: src/runtime_src/hip/core/memory_pool.cpp [hunk 16]
```diff
@@ -0,0 +1,407 @@
...
+
+    m_list.emplace(m_list.end(), std::make_shared<memory_pool_node>(m_device, m_pool_size, m_last_id++));
+  }
+
+  void
+  memory_pool::get_attribute(hipMemPoolAttr attr, void* value)
+  {
+    if (m_list.size() == 0) {
+      init();
+    }
```
Comment (stsoe) [inline src/runtime_src/hip/core/memory_pool.cpp:137]:
```suggestion
    if (m_list.empty())
      init();

```

File: src/runtime_src/hip/core/memory_pool.cpp [hunk 17]
```diff
@@ -0,0 +1,407 @@
...
+
+    switch (attr)
+    {
```
Comment (stsoe) [inline src/runtime_src/hip/core/memory_pool.cpp:140]:
```suggestion
    switch (attr) {
```

File: src/runtime_src/hip/core/memory_pool.cpp [hunk 18]
```diff
@@ -0,0 +1,407 @@
...
+      break;
+    };
+  }
+
+  void
+  memory_pool::set_attribute(hipMemPoolAttr attr, void* value)
+  {
+    if (m_list.size() == 0) {
+      init();
+    }
```
Comment (stsoe) [inline src/runtime_src/hip/core/memory_pool.cpp:180]:
```suggestion
    if (m_list.empty())
      init();
```

Same pattern repeated many times, so please fix all occurences.

File: src/runtime_src/hip/core/memory_pool.cpp [hunk 19]
```diff
@@ -0,0 +1,407 @@
...
+
+    size_t page_size = xrt_core::getpagesize();
+    size_t aligned_size = ((size + page_size) / page_size) * page_size;
+
+    std::lock_guard lock(m_mutex);
+
+    if (aligned_size > m_pool_size)
+    {
+      throw std::runtime_error("requested size is greater than memory pool block size.");
+    }
```
Comment (stsoe) [inline src/runtime_src/hip/core/memory_pool.cpp:245]:
```suggestion
    if (aligned_size > m_pool_size)
      throw std::runtime_error("requested size is greater than memory pool block size.");

```

File: src/runtime_src/hip/core/memory_pool.cpp [hunk 20]
```diff
@@ -0,0 +1,407 @@
...
+
+    std::shared_ptr<memory_pool_node> mm;
+
+    uint32_t num_tries = 0;
+    // find first free slot that fits, if none is found, enlarge the pool and try one more time
+    while (num_tries <= 2) {
+      ++ num_tries;
+      auto mp_itr = m_list.begin();
+      while (mp_itr != m_list.end())
+      {
```
Comment (stsoe) [inline src/runtime_src/hip/core/memory_pool.cpp:255]:
```suggestion
      while (mp_itr != m_list.end()) {
```
Consistency please,   Line 251 is good, why the newline here?  Can you fix all occurrences of newline after condition before curly brace?

File: src/runtime_src/hip/core/memory_pool.cpp [hunk 21]
```diff
@@ -0,0 +1,407 @@
...
+        mm = *mp_itr;
+        if (m_pool_size - mm->m_used < aligned_size)
+        {
+          mp_itr++;
+          continue;
+        }
+
+        auto _free = mm->m_free_list;
+        auto _not_free = _free;
```
Comment (stsoe) [inline src/runtime_src/hip/core/memory_pool.cpp:264]: Please never leading underscores.  Plus `not_free` makes no sense as it is assigned `free` ???

File: src/runtime_src/hip/core/memory_pool.cpp [hunk 22]
```diff
@@ -0,0 +1,407 @@
...
+    return true;
+  }
+
+  void memory_pool::malloc(void**ptr, size_t size)
+  {
+    *ptr = alloc(size);
+  }
+
+  void*
+  memory_pool::alloc(size_t size)
```
Comment (stsoe) [inline src/runtime_src/hip/core/memory_pool.cpp:231]: This function is much too long.  Anything longer than 15 lines require significant commenting.   Prefer refactoring so that code reads well even without comments.

File: src/runtime_src/hip/core/memory_pool.cpp [hunk 23]
```diff
@@ -0,0 +1,407 @@
...
+            size_t new_mem_start = reinterpret_cast<uint64_t>(mm->m_memory->get_address()) + _not_free->m_start;
+            return reinterpret_cast<void*>(new_mem_start);
+          }
+          _free = _free->m_next;
+        }
+        mp_itr++;
+        mm = *mp_itr;
+      }
+      
+      if (num_tries == 2) break;
```
Comment (stsoe) [inline src/runtime_src/hip/core/memory_pool.cpp:311]:
```suggestion
      if (num_tries == 2) 
        break;
```
Please fix all such occurences.

**Review by stsoe — CHANGES_REQUESTED** [review]
Many style changes requested.  Thanks for commenting.

File: src/runtime_src/hip/core/memory_pool.cpp (hunk 23 above)
Comment (zhangchiming) [reply src/runtime_src/hip/core/memory_pool.cpp:311]: Done.

File: src/runtime_src/hip/core/memory_pool.cpp (hunk 21 above)
Comment (zhangchiming) [reply src/runtime_src/hip/core/memory_pool.cpp:264]: underscores leading has been removed and variable name was changed from "not_free" to "alloc_slot", since it represent the slot that is chosen from current free slots for allocation.

File: src/runtime_src/hip/core/memory_pool.cpp (hunk 20 above)
Comment (zhangchiming) [reply src/runtime_src/hip/core/memory_pool.cpp:255]: Fixed.

File: src/runtime_src/hip/core/memory_pool.cpp (hunk 19 above)
Comment (zhangchiming) [reply src/runtime_src/hip/core/memory_pool.cpp:245]: Fixed.

File: src/runtime_src/hip/core/memory_pool.cpp (hunk 15 above)
Comment (zhangchiming) [reply src/runtime_src/hip/core/memory_pool.cpp:127]: Fixed.

File: src/runtime_src/hip/core/memory_pool.cpp (hunk 14 above)
Comment (zhangchiming) [reply src/runtime_src/hip/core/memory_pool.cpp:101]: Fixed.

File: src/runtime_src/hip/core/memory_pool.cpp (hunk 13 above)
Comment (zhangchiming) [reply src/runtime_src/hip/core/memory_pool.cpp:88]: fixed.

File: src/runtime_src/hip/core/memory_pool.cpp (hunk 12 above)
Comment (zhangchiming) [reply src/runtime_src/hip/core/memory_pool.cpp:91]: fixed.

File: src/runtime_src/hip/core/memory_pool.cpp (hunk 11 above)
Comment (zhangchiming) [reply src/runtime_src/hip/core/memory_pool.cpp:71]: fixed.

File: src/runtime_src/hip/core/memory_pool.cpp (hunk 10 above)
Comment (zhangchiming) [reply src/runtime_src/hip/core/memory_pool.cpp:48]: fixed.

File: src/runtime_src/hip/core/memory_pool.cpp (hunk 9 above)
Comment (zhangchiming) [reply src/runtime_src/hip/core/memory_pool.cpp:37]: Fixed.

File: src/runtime_src/hip/api/hip_memory.cpp (hunk 8 above)
Comment (zhangchiming) [reply src/runtime_src/hip/api/hip_memory.cpp:206]: fixed.

File: src/runtime_src/hip/core/event.cpp (hunk 7 above)
Comment (zhangchiming) [reply src/runtime_src/hip/core/event.cpp:2]: fixed.

File: src/runtime_src/hip/core/memory_pool.cpp (hunk 22 above)
Comment (zhangchiming) [reply src/runtime_src/hip/core/memory_pool.cpp:248]: This function is shortened and more comment has been added.

**Review by chvamshi-xilinx — APPROVED** [review]

**Comment by zhangchiming** [comment]
retest this please

**Comment by chvamshi-xilinx** [comment]
Merging this change. Will send separate PR later if there are any cleanups..
